Unit Tests for Enumerable.Select

MarcinJuraszek opened the pull request:
Set of basic tests for:

    IEnumerable<TResult> Enumerable.Select<TSource, TResult>(
        this IEnumerable<TSource> source,
        Func<TSource, TResult> selector)

1. Verify `ArgumentNullException`is thrown when either `source` or `selector` is null.
2. Verify that deferred execution works (for `T[]`, `List<T>` and `IEnumerable<T>`)
3. Verify that returned collection contains expected items (for `T[]`, `List<T>` and `IEnumerable<T>`)
4. Verify that `GetEnumerator` called twice returns two different instances of enumerator.

Tests are intended to improve code coverage around `Select` method before applying performance improvements to it (https://github.com/dotnet/corefx/pull/1282). It's also connected to https://github.com/dotnet/corefx/pull/1241 and https://github.com/dotnet/corefx/issues/1143.

Clockwork-Muse commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,230 @@
+            IEnumerable<int> source = null;
+            Func<int, int> selector = i => i + 1;
+
+            AssertException<ArgumentNullException>(() => source.Select(i => i + 1));
```
There's `Assert.Throws<T>(Action)`...

MarcinJuraszek replied:
Getting following error when trying to use it: 

> SelectTests.cs(15,13): error CS0012: The type '`Task`' is defined in an assembly
that is not referenced. You must add a reference to assembly 'System.Threading.
Tasks, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [d:\
GitHub\corefx\src\System.Linq\tests\System.Linq.Tests.csproj]

I didn't want to do that, but if you say it's ok, I'm fine with that.

Clockwork-Muse replied:
I didn't realize it had that dependency (or that this would manifest).  Probably because I'm dealing with the Parallel version, so the package was imported for other tests.

I think I prefer using the xunit-provided version, even with that caveat.  You should only need to add it as a project package - you don't have to list it as a `using`.  Although I don't know if I count as "on the team" for this purpose...

stephentoub replied:
Please feel free to add System.Threading.Tasks to the packages.config for System.Linq.Tests.  We're fairly strict about what gets added to packages.config for the product assemblies, but we're much more liberal with the tests.

MarcinJuraszek replied:
Will add a reference and change to use `Assert.Throws`.

ikopylov commented:
:+1: 

It would be great to have tests that verify that ```Select``` works in conjunction with several other LINQ methods: ```Where```, ```Contains```, ```ToArray```. Perhaps, it is better to add them in a separate PR.

stephentoub commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,230 @@
+        }
+
+        [Fact]
+        public void Select_SourceIsAnArray_ExecutionIsDefered()
```
Nit: "deferred" is spelled with two 'r's

stephentoub commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,230 @@
+            try
+            {
+                action();
+            }
```
As noted in other comments, it'd be good to use Assert.Throws instead of this custom version.  But on the off chance that you don't, you should add an Assert.False here after action(), to make sure that action does indeed throw an exception.

MarcinJuraszek replied:
AFAIK test will fail if no assert is executed, but yeah, I'm changing to `Assert.Throws` and adding *System.Threading.Tasks* reference.

stephentoub commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,230 @@
+            var query = source.Select(i => i + 1);
+
+            var enumerator1 = query.GetEnumerator();
+            var enumerator2 = query.GetEnumerator();
```
I'd suggest asserting here that query and enumerator1 are the same object reference.  That's an important optimization employed by the current implementation.

MarcinJuraszek replied:
Added the assert.

stephentoub commented:
@MarcinJuraszek, thanks for adding these!  Much appreciated, and they're a good start for Select tests.  In addition to these and the ones that @ikopylov mentions (i.e. Select in combination with other operators), I'd hope to also see tests for a variety of other things, such as:
- The other Select overload with a function that accepts an index in addition to the input state
- Various other sources, including an Empty source, a source with many elements (potentially > Int32.MaxValue, but that would run for a very long time and would likely need to be an outer loop test), an IReadOnlyCollection, an ICollection, etc.
- Exceptions being thrown from the selector, exceptions thrown from various methods on the source enumerable/enumerator
- Attempting to reuse the enumerator returned from Select, attempting to use the enumerator after done iterating, etc.

Clockwork-Muse commented:
@ikopylov - something like the [ParallelQueryCombinationTests](https://github.com/Clockwork-Muse/corefx/blob/516_System.Linq.Parallel/src/System.Linq.Parallel/tests/ParallelQueryCombinationTests.cs)?  Still not quite done, but...
There was something like this as part of the original tests, but I've been expanding it some.  (For one thing, I'm reasonably certain I can nest however many layers deep, now).

@stephentoub - how much of that last bit should I add to the PLINQ tests?  I didn't think about trying to reuse iterators... hmm..
I don't have anything that really goes over `int.MaxValue` elements, although I guess I'd have to make something special for that...

MarcinJuraszek commented:
Thanks for all the comments.

- Added `System.Threading.Tasks` reference and removed custom *does the code throws exception* code in favor of `Assert.Throws<TException>`.
- Added tests based on different sources (`Enumerable.Empty`, `IReadOnlyCollection<T>` and `ICollection<T>`).
- Test that uses a source collection with more elements than `int.MaxValue` takes ~80s to execute. Didn't include in the PR.
- Added tests where selector or enumerator throw exception from different methods/properties.

About other suggested tests:
- I deliberately didn't include tests that use other LINQ methods. I think they should be tested separately. I wanted to test `Select` and `Select` only and using other LINQ methods seems like the thing that's actually tests gets blurred between different operators. That's why I use the enumerator directly (or `foreach` loop in some cases) in tests included in that Pull Request. **Thoughts?**
- I do agree that there is value in tests around the other `Select` overload, however as this Pull Request is to get confidence that https://github.com/dotnet/corefx/pull/1282 doesn't break anything so I will not implement them here. They can be tracked by https://github.com/dotnet/corefx/issues/1143 or by other work that intends to refactor/improve `Select` overload that takes a function with index. **Maybe we need one item with a table and a list of checkboxes, one per LINQ method, to see what's being worked on, what's planned, what's up for grabs, etc?**
- What exactly should be tested for *Attempting to reuse the enumerator returned from Select, attempting to use the enumerator after done iterating, etc.*? I can write tests that make sure that after `MoveNext` returns false `Current` is set to `default(TResult)`. **Is there anything else that they should validate?**

@stephentoub Running code coverage doesn't work for me locally. Is it possible to get a code coverage report for a Pull Request?

stephentoub commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,590 @@
+        public void Select_SourceIsNull_ArgumentNullExceptionThrown()
+        {
+            IEnumerable<int> source = null;
+            Func<int, int> selector = i => i + 1;
```
```selector``` doesn't appear to be used.

MarcinJuraszek replied:
Fixed.

stephentoub commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,590 @@
+        public void Select_SourceIsIEnumerable_ReturnsExpectedValues()
+        {
+            int nbOfItems = 5;
+            IEnumerable<int> source = Enumerable.Range(1, 5);
```
Nit: since you define ``nbOfItem```, I'd expect the range to be ```Enumerable.Range(1, nbOfItems)``` rather than ```Enumerable.Range(1, 5)```.

stephentoub commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,590 @@
+        {
+            IEnumerable<int> source = Enumerable.Empty<int>();
+
+            IEnumerable<int> result = source.Select(i => i + 1);
```
Consider also setting a captured Boolean flag in the delegate and then asserting below that it's not getting set, in order to verify not only that Select didn't yield any results but also never executed the delegate.

stephentoub commented on src/System.Linq/tests/System.Linq.Tests.csproj:
```diff
@@ -36,5 +37,10 @@
       <Name>System.Linq</Name>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup>
```
This shouldn't be necessary... it should happen implicitly due to having added the package entry to packages.config.  In fact, I'm a bit surprised having it here as well isn't causing a warning.

MarcinJuraszek replied:
msbuild was happy with just entry in packages.config, but VS wasn't. I'll investigate.

stephentoub commented:
> Thanks for all the comments.

Thanks for addressing them!  The tests are shaping up nicely.

> Test that uses a source collection with more elements than int.MaxValue takes ~80s to execute. Didn't include in the PR.

That's fine.  If you wanted to (you don't need to), you could include it but mark it as ```[OuterLoop]```.  That way it'll only run when we run the "outer loop" tests, which we do a few times a day in the lab.

> I deliberately didn't include tests that use other LINQ methods. I think they should be tested separately. I wanted to test Select and Select only and using other LINQ methods seems like the thing that's actually tests gets blurred between different operators. That's why I use the enumerator directly (or foreach loop in some cases) in tests included in that Pull Request. Thoughts?

That's fair.  My concern, however, is that your other PR is changing what interfaces are implemented by the returned enumerable, which means that by design it does affect the interaction between operators... in an expression like ```Select(...).Count(...)```, your other PR will end up changing what code is executed.  And as such, we should have tests that verify the expected behavior of such situations.  Reasonable?  We should at least have a few tests that verify expected behavior when combined with operators that explicitly query for ```ICollection<T>``` and have special code paths for those cases.  From a cursory skim, I believe that would include Contains, Count, various To* methods, Reverse, and OrderBy.

> I do agree that there is value in tests around the other Select overload, however as this Pull Request is to get confidence that #1282 doesn't break anything so I will not implement them here.

Ok, so #1282 isn't changing the implementation of the other Select overload?  If so, sounds good.

> Maybe we need one item with a table and a list of checkboxes, one per LINQ method, to see what's being worked on, what's planned, what's up for grabs, etc?

I like that idea.

> What exactly should be tested for Attempting to reuse the enumerator returned from Select, attempting to use the enumerator after done iterating, etc.? I can write tests that make sure that after MoveNext returns false Current is set to default(TResult). Is there anything else that they should validate?

That's the kind of thing I was thinking.

> Running code coverage doesn't work for me locally.

How are you trying to run it, and what's the outcome?

MarcinJuraszek commented:
> That's fair. My concern, however, is that your other PR is changing what interfaces are implemented by the returned enumerable, which means that by design it does affect the interaction between operators... in an expression like `Select(...).Count(...)`, your other PR will end up changing what code is executed. And as such, we should have tests that verify the expected behavior of such situations. Reasonable? We should at least have a few tests that verify expected behavior when combined with operators that explicitly query for `ICollection<T>` and have special code paths for those cases. From a cursory skim, I believe that would include `Contains`, `Count`, various To* methods, `Reverse`, and `OrderBy`.

I would rather add test that verifies thats `Select` called on `T[]`, `List<T>` or `ICollection<T>` correctly propagates `Count`/`Length` from source collection, and another set of tests to verify that `Count()`, `Reverse()` and other methods that use `ICollection<T>.Count` use it correctly. This way we can test the methods separately and make sure that each of the components work as expected, without having tests that span multiple operators (which kind of break the rule to test one thing at a time).

Ofc that new `Select` test can only be added as part of https://github.com/dotnet/corefx/pull/1282 because it would fail with old implementation when `Select` returns iterator that doesn't implement `ICollection<T>`.

> Ok, so https://github.com/dotnet/corefx/pull/1282 isn't changing the implementation of the other Select overload? If so, sounds good.

You're correct, https://github.com/dotnet/corefx/pull/1282 does not change how the other `Select` overload works. I guess there is space for improvements there as well, but for now I'm focused on `Select()` that takes `IEnumerable<TSource>` and `Func<TSource, TResult>`.

> That's the kind of thing I was thinking.

I'll add those later tonight.

> How are you trying to run it, and what's the outcome?

I've run `msbuild /p:Coverage=true` but it returned with `1` and only provides coverage report for couple libraries, and `System.Linq` is not one of them. I'll rerun it later tonight and include error message if I can find one.

MarcinJuraszek commented:
Oh, and about that:

> >  Maybe we need one item with a table and a list of checkboxes, one per LINQ method, to see what's being worked on, what's planned, what's up for grabs, etc?

> I like that idea.

I think someone from the team should own that Issue, so it's easier to keep updated (imo it's much less likely that someone leaves the team than that someone got bored as OSS contributor :))

MarcinJuraszek commented:
After running `build.cmd /p:Coverage=true` result file for `System.Linq.Tests` is just 1KB with following content:

    <?xml version="1.0" encoding="utf-8"?>
    <CoverageSession xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
          <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="0" />
          <Modules />
    </CoverageSession>

stephentoub commented:
> I would rather add test that verifies that  Select  called on  T[] ,  List<T>  or  ICollection<T>  correctly propagates  Count / Length  from source collection, and another set of tests to verify that  Count() ,  Reverse()  and other methods that use  ICollection<T>.Count  use it correctly. This way we can test the methods separately and make sure that each of the components work as expected, without having tests that span multiple operators (which kind of breaks the rule to test one thing at a time).

In your other PR, you added an implementation of ```ICollection<T>``` to the Select iterator, and that implementation would end up breaking subsequent consumers due to throwing an exception from Contains.  This suggestion would not detect such breaks.

> After running  build.cmd /p:Coverage=true  result file for  System.Linq.Tests 

Can you try doing a ```git clean -xdf``` from your repo's root and then doing ```build.cmd /p:Coverage=true``` also from the root?  I'm wondering if something got corrupted along the way.

eatdrinksleepcode commented:
I am wondering if this is the result of an incremental build. Had you already done a successful test run prior to running for coverage? I am thinking we need to add the coverage report file to the test output when coverage is requested.

> On Apr 2, 2015, at 12:23 AM, Marcin Juraszek <notifications@github.com> wrote:
> 
> After running build.cmd /p:Coverage=true result file for System.Linq.Tests is just 1KB with following content:
> 
> <?xml version="1.0" encoding="utf-8"?>
> <CoverageSession xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
>       <Summary numSequencePoints="0" visitedSequencePoints="0" numBranchPoints="0" visitedBranchPoints="0" sequenceCoverage="0" branchCoverage="0" maxCyclomaticComplexity="0" minCyclomaticComplexity="0" />
>       <Modules />
> </CoverageSession>
> —
> Reply to this email directly or view it on GitHub.
>

MarcinJuraszek commented:
> Can you try doing a `git clean -xdf` from your repo's root and then doing `build.cmd /p:Coverage=true` also from the root? I'm wondering if something got corrupted along the way.

Still no luck.

stephentoub commented:
LGTM

stephentoub commented:
@MarcinJuraszek, thanks for adding these.  Can you rebase so that this is mergable?

hpshelton commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,662 @@
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsIEnumerable_ExecutionIsDefered()
```
Spelling: "Deferred"

hpshelton commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,662 @@
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsAList_ExecutionIsDefered()
```
Spelling: "Deferred"

hpshelton commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,662 @@
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsAnArray_ExecutionIsDefered()
```
Spelling: "Deferred"

hpshelton commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,662 @@
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(index);
```
I personally like relying on the values in `source` rather than on the index so someone could safely make modifications to the array without changing the loop structure.

hpshelton commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,662 @@
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(index);
```
I personally like relying on the values in `source` rather than on the index so someone could safely make modifications to the array without changing the loop structure.

hpshelton commented on src/System.Linq/tests/SelectTests.cs:
```diff
@@ -0,0 +1,662 @@
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(index);
```
I personally like relying on the values in `source` rather than on the index so someone could safely make modifications to the array without changing the loop structure.

MarcinJuraszek replied:
There is no array here.

MarcinJuraszek commented:
Trying to merge and rebase now. Don't have much git experience, so hopefully it end up well :)

MarcinJuraszek commented:
OK, didn't work well. Will do it tomorrow with someone who knows git.

hpshelton commented:
@eatdrinksleepcode - the incremental build won't re-run code coverage if the binaries haven't changed and the `coverage` output directory exists (this may be by design). However, even when tests are run, coverage reports aren't generated for System.Linq.dll in particular. I opened https://github.com/dotnet/corefx/issues/1479 with more info.

theoy commented:
/cc @VSadov

stephentoub commented:
@MarcinJuraszek, just checking in on this.  Have you had any luck fixing the rebase?

VSadov commented:
Re: someone who knows git.
I found that using TortoiseGit helped a lot with things like this.
I think it is the command line interface that makes Git look more complex than it really is. A tool with a visual feedback makes it much easier to figure what is going on.

mjrousos commented:
Another option for coming up to speed on git would be to use [SourceTree](https://www.atlassian.com/software/sourcetree/overview). It's a free git GUI that can help you try out various operations without having to know all of the command-line syntax (and in most cases you can see the command line executed, so you can learn how to do the same things directly from the command prompt).

MarcinJuraszek commented:
Sorry, didn't have much time to do anything lately. Will try finishing it up tomorrow/Tuesday so it can be merged.

MarcinJuraszek commented:
@stephentoub , hope that's what you were talking about.

stephentoub commented:
Thanks for the fixes, @MarcinJuraszek, and of course for the tests in general.  Looks good.

stephentoub commented:
@VSadov, look ok?

VSadov commented:
LGTM thanks!!
